File: asg4.pl
% CSCI3180 Principles of Programming Languages

% --- Declaration ---

% I declare that the assignment here submitted is original except for source
% material explicitly acknowledged. I also acknowledge that I am aware of
% University policy and regulations on honesty in academic work, and of the
% disciplinary guidelines and procedures applicable to breaches of such policy
% and regulations, as contained in the website
% http://www.cuhk.edu.hk/policy/academichonesty/
% Credits to the my roommate Huzeyfe KIRAN(1155104019).
% Assignment 4
% Name : Alshir Soyunjov
% Student ID : 1155119170

% Rules
sum(0,X,X).
sum(s(X),Y,s(Z)) :- sum(X,Y,Z).

% 1.a Rules
natNum(0).
natNum(s(X)) :- natNum(X).

% 1.b Rules
lt(X,s(X)).
lt(X,Y) :- sum(X,s(Z),Y), natNum(Z).
% % ?- lt(X,s(s(s(0)))).
% ?- natNum(lt(X,3)).

% 1.d Rules
geq(X,Y) :- sum(Z,Y,X), natNum(Z).

% 1.e Rules
max(X,Y,Z) :- geq(Z,X), geq(Z,Y).

% 1.f Rules
difference(X,Y,Z) :- sum(Y,Z,X).

% 1.g Rules
mod(X,Y,Z) :- lt(Z,Y), geq(X,Z), difference(X,Z,D), multiple(Y,D).
multiple(X,0) :- X=s(_).
multiple(X,Y) :- X=s(_), Y=s(_), geq(Y,X), difference(Y,X,D), multiple(X,D).

% 2.a Rules
% bt(a,bt(b,nil,bt(d,nil,nil)),bt(c,bt(e,bt(f,nil,nil),bt(g,nil,nil)),nil)).

% 2.b Rules
isTree(nil).
isTree(bt(_,L,R)) :- isTree(L), isTree(R).

% 2.c Rules
isLeaf(X,bt(X,nil,nil)).
isLeaf(X,bt(Rt,L,R)) :- isLeaf(X,L).
isLeaf(X,bt(Rt,L,R)) :- isLeaf(X,R).

% 2.d Rules
numberOfLeaf(nil,N) :- N=0.
numberOfLeaf(bt(E,nil,nil),N) :- isTree(bt(E,nil,nil)), N=s(0).
numberOfLeaf(bt(Rt,L,R),N) :- isTree(bt(Rt,L,R)), numberOfLeaf(L,LN), numberOfLeaf(R,RN), sum(LN,RN,N).

% 2.e Rules	
height(nil, 0).  
height(bt(_,L,R),H) :- height(L,L1), height(R,R1), max(R1,L1,M), sum(s(0),M,H). 

File: asg4.ml
(* CSCI3180 Principles of Programming Languages

* --- Declaration ---

* I declare that the assignment here submitted is original except for source
* material explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
* Credits to the my roommate Huzeyfe KIRAN(1155104019).
* Assignment 4
* Name : Alshir Soyunjov
* Student ID : 1155119170
*)
datatype suit = Clubs | Diamonds | Hearts | Spades;
datatype rank = Jack | Queen | King | Ace | Num of int;
type card = suit * rank;

datatype color = Red | Black;
datatype move = Discard of card | Draw;

fun abs x = if x >= 0 then x else ~x;

fun cardColor(x:suit,y:rank):color = if x=Diamonds orelse x=Hearts then Red else Black;

fun cardValue(_,Num i) = i
|   cardValue(_,Jack) = 10
|   cardValue(_,Queen) = 10
|   cardValue(_,King) = 10
|   cardValue(_,Ace) = 11;

fun numOfCard(cardList:(suit*rank) list): int =  
         if null(cardList) then 0
         else 1 + numOfCard(tl(cardList));

fun removeCard(l:(suit*rank) list, s:suit, r:rank):(suit*rank) list =
        if (s,r) = hd(l) then tl(l)
        else [hd(l)]@removeCard(tl(l),s,r);

fun sumCards(cardList:(suit*rank) list, col:color):int =
        if null(cardList) then 0
        else if cardColor(hd(cardList)) = col then cardValue(hd(cardList)) + sumCards(tl(cardList),col)
        else sumCards(tl(cardList),col);

fun score(cardList:(suit*rank) list):int = abs(sumCards(cardList,Red) - sumCards(cardList,Black)) + 6 - numOfCard(cardList);

fun runGame(cardList,moveList) =
            let
                    fun runGameA([],Draw,moveList,heldCards) = score(heldCards)
                    |   runGameA(cardList,Draw,moveList,heldCards) =
                                if length(hd(cardList)::heldCards) > 5 orelse null(moveList) then score(hd(cardList)::heldCards)
                                else runGameA(tl(cardList),hd(moveList),tl(moveList),hd(cardList)::heldCards)
                    |   runGameA(cardList,Discard (x,y),moveList,heldCards) =
                                if null(moveList) then score(removeCard(heldCards,x,y))
                                else runGameA(cardList,hd(moveList),tl(moveList),removeCard(heldCards,x,y))

            in
                    runGameA(cardList,hd(moveList),tl(moveList),[])
            end;



